org.eclipse.m2e.core.ui/src/org/eclipse/m2e/core/ui/internal/archetype/ArchetypeGenerator.java

```diff
@@ -122,6 +123,12 @@ public Collection<MavenProjectInfo> createArchetypeProjects(IPath location, IArc
           }
         }
         Thread.onSpinWait();
+        // honor cancellations by informing our Future about it then exit 
```

Comment:
I suggest to call just `subMonitor.checkCanceled();` here and above in the body of the while-loop that calls `readAndDispatch()`.
Then `maven.cancel(true);`  should be called in the catch-clause for `CancellationExceptions`. In order to have the maven variable available there the code that assigned that variable has to be moved before the try block.

Reply:
I tried to keep a minimal change.

As I look further, I kind of find little things that would need pulling on.

AbstractProjectScanner has a run declared to throw InterruptedException
use of LocalProjectScanner here catches that and returns an empty list.
What is the intent of that, vs bailing out the way the monitor cancellation is actually handled in this code? (throwing an exception with a Cancel status). 

I would like to propose sorting out AbstractProjectScanner, removing the declaration of InterruptedException, (since it has a monitor it does cancellation checks already that way), then handling OperationCancelledException for cancellation... except that this special return of an empty list stands in the way of rationalizing this.

Comments on how to proceed? Do you recognize the advantage or use of an empty list return vs throwing with exception? 

Our product uses this ArchetypeGenerator
We obviously already handle CoreException
Common ways of handling a CoreException via ErrorDialog suppress CANCEL which is also good (the user asked for the cancel usually, don't tell them)

so at least in our use case this would be the way to go - all OperationCancelledExceptions turn into a CoreException with status CANCEL

Reply:
see resulting proposal with https://github.com/eclipse-m2e/m2e-core/pull/955/commits/99bddd0288db18ddb2312cfb0d19da9990b231f1